File: crates/enchu-extend/bench_vs_meilisearch.ts
/**
 * Enchu Extend vs Meilisearch — マルチテナント ファセット検索バトル
 *
 * 同一データ100万件で検索速度を比較。
 * Enchu: napi経由（queryFast/sliceFast）
 * Meilisearch: HTTP API
 */

const enchu = require('./enchu-extend.node')

const PG_CONN = 'postgresql://postgres:enchu@localhost:5433/enchu_bench'
const CACHE_DIR = '/tmp/enchu_vs_meili'
const MEILI_URL = 'http://localhost:7700'
const MEILI_KEY = 'bench123'

const TENANTS = 100
const PRODUCTS_PER_TENANT = 10_000
const N = TENANTS * PRODUCTS_PER_TENANT

const CATEGORIES = 20
const COLORS = 10
const SIZES = 5

async function main() {
  console.log('╔══════════════════════════════════════════════════════════════════════╗')
  console.log(`║  Enchu Extend vs Meilisearch — マルチテナント ファセット検索       ║`)
  console.log(`║  ${N.toLocaleString()} 件 (${TENANTS}テナント × ${PRODUCTS_PER_TENANT.toLocaleString()}商品)                          ║`)
  console.log('╚══════════════════════════════════════════════════════════════════════╝\n')

  // ══════ Pg にデータ投入 ══════
  const { Client } = await import('pg')
  const pg = new Client({ connectionString: PG_CONN })
  await pg.connect()

  console.log('  [1] Pg セットアップ...')
  const t1 = Date.now()
  await pg.query('DROP TABLE IF EXISTS products CASCADE')
  await pg.query(`
    CREATE TABLE products (
      id SERIAL PRIMARY KEY,
      tenant_id INT NOT NULL,
      category INT NOT NULL,
      color INT NOT NULL,
      size INT NOT NULL,
      price INT NOT NULL
    )
  `)
  const BATCH = 1000
  for (let i = 0; i < N; i += BATCH) {
    const values: string[] = []
    for (let j = 0; j < BATCH && i + j < N; j++) {
      const id = i + j
      const tenant = Math.floor(id / PRODUCTS_PER_TENANT)
      const p = id % PRODUCTS_PER_TENANT
      values.push(`(${tenant}, ${p % CATEGORIES}, ${p % COLORS}, ${p % SIZES}, ${p % 10000})`)
    }
    await pg.query(`INSERT INTO products (tenant_id, category, color, size, price) VALUES ${values.join(',')}`)
  }
  console.log(`      ${N.toLocaleString()} 件 (${((Date.now() - t1) / 1000).toFixed(1)}s)`)

  // ══════ Enchu セットアップ ══════
  console.log('\n  [2] Enchu セットアップ...')
  const t2 = Date.now()
  const fs = await import('fs')
  fs.rmSync(CACHE_DIR, { recursive: true, force: true })

  enchu.autoInit(CACHE_DIR, PG_CONN)

  // Pgからロード
  const LOAD_BATCH = 10000
  let offset = 0
  while (offset < N) {
    const res = await pg.query(
      `SELECT id, tenant_id, category, color, size, price FROM products ORDER BY id LIMIT $1 OFFSET $2`,
      [LOAD_BATCH, offset]
    )
    if (res.rows.length === 0) break
    for (const row of res.rows) {
      const eid = enchu.insert('products')
      enchu.writeInt('products', eid, 'tenant_id', row.tenant_id)
      enchu.writeInt('products', eid, 'category', row.category)
      enchu.writeInt('products', eid, 'color', row.color)
      enchu.writeInt('products', eid, 'size', row.size)
      enchu.writeInt('products', eid, 'price', row.price)
    }
    offset += res.rows.length
  }
  enchu.flush()
  console.log(`      Enchu: ${enchu.count('products')} 件 (${((Date.now() - t2) / 1000).toFixed(1)}s)`)

  // ID事前解決
  const tid = enchu.resolveTable('products')
  const fTenant = enchu.resolveField('products', 'tenant_id')
  const fCat = enchu.resolveField('products', 'category')
  const fColor = enchu.resolveField('products', 'color')
  const fSize = enchu.resolveField('products', 'size')

  // ══════ Meilisearch セットアップ ══════
  console.log('\n  [3] Meilisearch セットアップ...')
  const t3 = Date.now()

  await meiliFetch('DELETE', '/indexes/products')
  await sleep(1000)
  await meiliFetch('POST', '/indexes', { uid: 'products', primaryKey: 'id' })
  await sleep(1000)
  await meiliFetch('PUT', '/indexes/products/settings/filterable-attributes',
    ['tenant_id', 'category', 'color', 'size', 'price'])
  await sleep(1000)

  // バッチ投入
  for (let tenant = 0; tenant < TENANTS; tenant++) {
    const docs: any[] = []
    for (let p = 0; p < PRODUCTS_PER_TENANT; p++) {
      const id = tenant * PRODUCTS_PER_TENANT + p
      docs.push({
        id,
        tenant_id: tenant,
        category: p % CATEGORIES,
        color: p % COLORS,
        size: p % SIZES,
        price: p % 10000,
      })
    }
    await meiliFetch('POST', '/indexes/products/documents', docs)
  }

  // indexing完了待ち
  process.stdout.write('      indexing...')
  while (true) {
    const resp = await meiliFetch('GET', '/tasks?statuses=enqueued,processing&limit=1')
    if (resp.total === 0) break
    await sleep(2000)
  }
  console.log(` (${((Date.now() - t3) / 1000).toFixed(1)}s)`)

  // ══════ バトル ══════
  const w = 50
  const r = 500

  console.log('\n  ┌───────────────────────────┬──────────────┬──────────────┬──────────┐')
  console.log('  │ クエリ                    │   Enchu (ns) │  Meili (ns)  │     倍率 │')
  console.log('  ├───────────────────────────┼──────────────┼──────────────┼──────────┤')

  // 1. 単条件: tenant_id = 42
  const en1 = benchSync(w, r, () => {
    return enchu.sliceFast(tid, fTenant, 42).length / 4
  })
  const me1 = await benchAsync(w, r, 'tenant_id = 42')
  row('tenant=42 (1cond)      ', en1, me1)

  // 2. 2条件: tenant_id=42 AND category=5
  const en2 = benchSync(w, r, () => {
    return enchu.queryCount(tid, [[fTenant, 42], [fCat, 5]])
  })
  const me2 = await benchAsync(w, r, 'tenant_id = 42 AND category = 5')
  row('tenant+cat (2cond)     ', en2, me2)

  // 3. 3条件: tenant_id=42 AND category=5 AND color=3
  const en3 = benchSync(w, r, () => {
    return enchu.queryCount(tid, [[fTenant, 42], [fCat, 5], [fColor, 3]])
  })
  const me3 = await benchAsync(w, r, 'tenant_id = 42 AND category = 5 AND color = 3')
  row('tenant+cat+col (3cond) ', en3, me3)

  // 4. 4条件
  const en4 = benchSync(w, r, () => {
    return enchu.queryCount(tid, [[fTenant, 42], [fCat, 5], [fColor, 3], [fSize, 1]])
  })
  const me4 = await benchAsync(w, r, 'tenant_id = 42 AND category = 5 AND color = 3 AND size = 1')
  row('tenant+4facets (4cond) ', en4, me4)

  // 5. 2条件でエンティティ取得（queryFast = 実データ返却）
  const en5 = benchSync(w, r, () => {
    return enchu.queryFast(tid, [[fTenant, 42], [fCat, 5]]).length / 4
  })
  const me5 = await benchAsync(w, r, 'tenant_id = 42 AND category = 5')
  row('tenant+cat (entities) ', en5, me5)

  console.log('  └───────────────────────────┴──────────────┴──────────────┴──────────┘')

  // cleanup
  await pg.query('DROP TABLE IF EXISTS products')
  await pg.end()
  await meiliFetch('DELETE', '/indexes/products')
  fs.rmSync(CACHE_DIR, { recursive: true, force: true })
  console.log('\n  Done.')
}

// ════════════════ helpers ════════════════

function benchSync(warmup: number, iterations: number, fn: () => number): number {
  for (let i = 0; i < warmup; i++) fn()
  const start = process.hrtime.bigint()
  for (let i = 0; i < iterations; i++) fn()
  return Number(process.hrtime.bigint() - start) / iterations
}

async function benchAsync(warmup: number, iterations: number, filter: string): Promise<number> {
  for (let i = 0; i < warmup; i++) await meiliSearch(filter)
  const start = process.hrtime.bigint()
  for (let i = 0; i < iterations; i++) await meiliSearch(filter)
  return Number(process.hrtime.bigint() - start) / iterations
}

async function meiliSearch(filter: string): Promise<number> {
  const resp = await meiliFetch('POST', '/indexes/products/search', {
    filter,
    limit: 10000,
    q: '',
  })
  return resp.estimatedTotalHits ?? 0
}

async function meiliFetch(method: string, path: string, body?: any): Promise<any> {
  const opts: RequestInit = {
    method,
    headers: {
      'Authorization': `Bearer ${MEILI_KEY}`,
      'Content-Type': 'application/json',
    },
  }
  if (body) opts.body = JSON.stringify(body)
  const resp = await fetch(`${MEILI_URL}${path}`, opts)
  if (!resp.ok && resp.status !== 404) {
    const text = await resp.text()
    throw new Error(`Meili ${method} ${path}: ${resp.status} ${text}`)
  }
  const text = await resp.text()
  return text ? JSON.parse(text) : {}
}

function sleep(ms: number) { return new Promise(r => setTimeout(r, ms)) }

function fmt(ns: number): string {
  if (ns >= 1_000_000) return `${(ns / 1_000_000).toFixed(1)}ms`
  if (ns >= 1_000) return `${(ns / 1_000).toFixed(1)}µs`
  return `${Math.round(ns)}ns`
}

function row(label: string, en: number, me: number) {
  const ratio = Math.floor(me / en)
  console.log(`  │ ${label} │ ${fmt(en).padStart(12)} │ ${fmt(me).padStart(12)} │ ${String(ratio + 'x').padStart(8)} │`)
}

main().catch(console.error)

File: crates/enchu-extend/bench/bench_vs_meilisearch.rs
//! Enchu v20 vs Meilisearch — マルチテナント ファセット検索ベンチ
//!
//! シナリオ: ECサイト、100テナント × 10,000商品 = 1,000,000件
//! 各商品: tenant_id, category(20種), color(10色), size(5種), price(0-9999)
//!
//! Meilisearch は HTTP サーバーとして起動済みが前提:
//!   meilisearch --master-key bench123 --db-path /tmp/meili_bench
//!
//! v20 Cylinder = prefix sum O(1) スライス。交差もゼロアロケーション。

use std::hint::black_box;
use std::time::Instant;

use enchu_v15::v20::engine::*;

const TENANTS: u32 = 100;
const PRODUCTS_PER_TENANT: u32 = 10_000;
const N: u32 = TENANTS * PRODUCTS_PER_TENANT; // 1,000,000

const CATEGORIES: u32 = 20;
const COLORS: u32 = 10;
const SIZES: u32 = 5;
const MAX_PRICE: u32 = 10_000;

const MEILI_URL: &str = "http://localhost:7700";
const MEILI_KEY: &str = "bench123";

fn main() {
    let rt = tokio::runtime::Runtime::new().unwrap();
    rt.block_on(async { run().await });
}

async fn run() {
    println!("╔══════════════════════════════════════════════════════════════════╗");
    println!("║  Enchu v20 vs Meilisearch — マルチテナント ファセット検索     ║");
    println!("║  {} テナント × {} 商品 = {} 件                       ║", TENANTS, PRODUCTS_PER_TENANT, N);
    println!("╚══════════════════════════════════════════════════════════════════╝\n");

    // ══════ Meilisearch check ══════
    print!("  Meilisearch 接続確認...");
    let client = reqwest::Client::new();
    match client.get(format!("{MEILI_URL}/health")).send().await {
        Ok(r) if r.status().is_success() => println!(" OK"),
        _ => {
            println!(" NG");
            println!("\n  ⚠  Meilisearch が起動していません。先に以下を実行:");
            println!("     meilisearch --master-key bench123 --db-path /tmp/meili_bench");
            return;
        }
    }

    // ══════ Enchu v20 setup ══════
    print!("  Enchu v20 setup...");
    let t = Instant::now();
    let enchu_dir = "/tmp/bench_vs_meili_enchu_v20";
    let _ = std::fs::remove_dir_all(enchu_dir);

    let schema = Schema::new("shop",
        TableDef::leaf("products", vec![
            ("tenant_id", FieldType::Int),
            ("category", FieldType::Int),
            ("color", FieldType::Int),
            ("size", FieldType::Int),
            ("price", FieldType::Int),
        ]),
    );
    let mut enchu = Engine::create(enchu_dir, schema).unwrap();

    for tenant in 0..TENANTS {
        for p in 0..PRODUCTS_PER_TENANT {
            let eid = enchu.insert("products").unwrap();
            enchu.write_i64("products", eid, "tenant_id", tenant as i64).unwrap();
            enchu.write_i64("products", eid, "category", (p % CATEGORIES) as i64).unwrap();
            enchu.write_i64("products", eid, "color", (p % COLORS) as i64).unwrap();
            enchu.write_i64("products", eid, "size", (p % SIZES) as i64).unwrap();
            enchu.write_i64("products", eid, "price", (p % MAX_PRICE) as i64).unwrap();
        }
    }
    enchu.flush().unwrap();
    println!(" {:.1}s", t.elapsed().as_secs_f64());

    // ID事前解決（ホットループ用）
    let tid = enchu.resolve_table("products");
    let f_tenant = enchu.resolve_field("products", "tenant_id");
    let f_cat = enchu.resolve_field("products", "category");
    let f_color = enchu.resolve_field("products", "color");
    let f_size = enchu.resolve_field("products", "size");
    let f_price = enchu.resolve_field("products", "price");

    // ══════ Meilisearch setup ══════
    print!("  Meilisearch setup...");
    let t = Instant::now();

    let _ = client.delete(format!("{MEILI_URL}/indexes/products"))
        .header("Authorization", format!("Bearer {MEILI_KEY}"))
        .send().await;
    tokio::time::sleep(std::time::Duration::from_secs(1)).await;

    client.post(format!("{MEILI_URL}/indexes"))
        .header("Authorization", format!("Bearer {MEILI_KEY}"))
        .json(&serde_json::json!({"uid": "products", "primaryKey": "id"}))
        .send().await.unwrap();
    tokio::time::sleep(std::time::Duration::from_secs(1)).await;

    client.put(format!("{MEILI_URL}/indexes/products/settings/filterable-attributes"))
        .header("Authorization", format!("Bearer {MEILI_KEY}"))
        .json(&serde_json::json!(["tenant_id", "category", "color", "size", "price"]))
        .send().await.unwrap();
    tokio::time::sleep(std::time::Duration::from_secs(1)).await;

    let mut id = 0u32;
    for tenant in 0..TENANTS {
        let mut docs = Vec::with_capacity(PRODUCTS_PER_TENANT as usize);
        for p in 0..PRODUCTS_PER_TENANT {
            docs.push(serde_json::json!({
                "id": id,
                "tenant_id": tenant,
                "category": p % CATEGORIES,
                "color": p % COLORS,
                "size": p % SIZES,
                "price": p % MAX_PRICE,
            }));
            id += 1;
        }
        client.post(format!("{MEILI_URL}/indexes/products/documents"))
            .header("Authorization", format!("Bearer {MEILI_KEY}"))
            .header("Content-Type", "application/json")
            .json(&docs)
            .send().await.unwrap();
    }

    print!(" indexing...");
    loop {
        let resp = client.get(format!("{MEILI_URL}/tasks?statuses=enqueued,processing&limit=1"))
            .header("Authorization", format!("Bearer {MEILI_KEY}"))
            .send().await.unwrap()
            .json::<serde_json::Value>().await.unwrap();
        if resp["total"].as_u64().unwrap_or(0) == 0 { break; }
        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
    }
    println!(" {:.1}s", t.elapsed().as_secs_f64());

    // ══════ Benchmark ══════
    let w = 50;
    let r = 1_000;

    println!("\n  ┌───────────────────────────┬──────────────┬──────────────┬──────────┐");
    println!("  │ クエリ                    │    Enchu v20 │  Meilisearch │     倍率 │");
    println!("  ├───────────────────────────┼──────────────┼──────────────┼──────────┤");

    // 1. 単条件: tenant_id = 42
    let en = bench(w, r, || {
        black_box(enchu.slice_by_id(tid, f_tenant, 42).len());
    });
    let me = bench_async(w, r, &client, "tenant_id = 42").await;
    row("tenant=42 (1cond)      ", en, me);

    // 2. 2条件: tenant_id=42 AND category=5
    let en = bench(w, r, || {
        let a = enchu.slice_by_id(tid, f_tenant, 42);
        let b = enchu.slice_by_id(tid, f_cat, 5);
        black_box(sorted_intersect_count(a, b));
    });
    let me = bench_async(w, r, &client, "tenant_id = 42 AND category = 5").await;
    row("tenant+cat (2cond)     ", en, me);

    // 3. 3条件: tenant_id=42 AND category=5 AND color=3
    let en = bench(w, r, || {
        let a = enchu.slice_by_id(tid, f_tenant, 42);
        let b = enchu.slice_by_id(tid, f_cat, 5);
        let c = enchu.slice_by_id(tid, f_color, 3);
        let ab = galloping_intersect(a, b);
        black_box(galloping_intersect_count(&ab, c));
    });
    let me = bench_async(w, r, &client, "tenant_id = 42 AND category = 5 AND color = 3").await;
    row("tenant+cat+col (3cond) ", en, me);

    // 4. 4条件: tenant_id=42 AND category=5 AND color=3 AND size=1
    let en = bench(w, r, || {
        let a = enchu.slice_by_id(tid, f_tenant, 42);
        let b = enchu.slice_by_id(tid, f_cat, 5);
        let c = enchu.slice_by_id(tid, f_color, 3);
        let d = enchu.slice_by_id(tid, f_size, 1);
        let ab = sorted_intersect(a, b);
        let abc = sorted_intersect(&ab, c);
        black_box(sorted_intersect_count(&abc, d));
    });
    let me = bench_async(w, r, &client, "tenant_id = 42 AND category = 5 AND color = 3 AND size = 1").await;
    row("tenant+3facets (4cond) ", en, me);

    // 5. range: tenant_id=42 AND price < 1000
    let en = bench(w, r, || {
        let a = enchu.slice_by_id(tid, f_tenant, 42);
        let b = enchu.slice_range("products", "price", 0, 999);
        black_box(sorted_intersect_count(a, b));
    });
    let me = bench_async(w, r, &client, "tenant_id = 42 AND price < 1000").await;
    row("tenant+price<1K (range)", en, me);

    println!("  └───────────────────────────┴──────────────┴──────────────┴──────────┘");

    let _ = std::fs::remove_dir_all(enchu_dir);
    let _ = client.delete(format!("{MEILI_URL}/indexes/products"))
        .header("Authorization", format!("Bearer {MEILI_KEY}"))
        .send().await;
    println!("\n  Done.");
}

fn bench<F: FnMut()>(w: usize, r: usize, mut f: F) -> f64 {
    for _ in 0..w { f(); }
    let s = Instant::now();
    for _ in 0..r { f(); }
    s.elapsed().as_nanos() as f64 / r as f64
}

async fn bench_async(w: usize, r: usize, client: &reqwest::Client, filter: &str) -> f64 {
    for _ in 0..w { meili_search(client, filter).await; }
    let s = Instant::now();
    for _ in 0..r { meili_search(client, filter).await; }
    s.elapsed().as_nanos() as f64 / r as f64
}

async fn meili_search(client: &reqwest::Client, filter: &str) {
    let resp = client.post(format!("{MEILI_URL}/indexes/products/search"))
        .header("Authorization", format!("Bearer {MEILI_KEY}"))
        .json(&serde_json::json!({
            "filter": filter,
            "limit": 10000,
            "q": ""
        }))
        .send().await.unwrap();
    let body = resp.json::<serde_json::Value>().await.unwrap();
    black_box(body["estimatedTotalHits"].as_u64().unwrap_or(0));
}

// ════════════════ ギャロッピング交差 ════════════════

#[inline]
fn gallop_ge(big: &[u32], val: u32, lo: usize) -> usize {
    let n = big.len();
    if lo >= n { return n; }
    if big[lo] >= val { return lo; }
    let mut step = 1usize;
    let mut hi = lo + step;
    while hi < n && big[hi] < val { step *= 2; hi = (lo + step).min(n); }
    let from = lo + step / 2;
    let to = hi.min(n);
    from + big[from..to].partition_point(|&x| x < val)
}

fn galloping_intersect(a: &[u32], b: &[u32]) -> Vec<u32> {
    let (small, big) = if a.len() <= b.len() { (a, b) } else { (b, a) };
    if small.is_empty() { return vec![]; }
    let mut result = Vec::with_capacity(small.len());
    let mut lo = 0usize;
    for &val in small {
        lo = gallop_ge(big, val, lo);
        if lo >= big.len() { break; }
        if big[lo] == val { result.push(val); lo += 1; }
    }
    result
}

fn galloping_intersect_count(a: &[u32], b: &[u32]) -> u32 {
    let (small, big) = if a.len() <= b.len() { (a, b) } else { (b, a) };
    if small.is_empty() { return 0; }
    let mut count = 0u32;
    let mut lo = 0usize;
    for &val in small {
        lo = gallop_ge(big, val, lo);
        if lo >= big.len() { break; }
        if big[lo] == val { count += 1; lo += 1; }
    }
    count
}

fn fmt(ns: f64) -> String {
    if ns >= 1_000_000.0 { format!("{:.1}ms", ns / 1_000_000.0) }
    else if ns >= 1_000.0 { format!("{:.1}µs", ns / 1_000.0) }
    else { format!("{:.0}ns", ns) }
}

fn row(label: &str, en: f64, me: f64) {
    let ratio = me / en;
    println!("  │ {} │ {:>12} │ {:>12} │ {:>6.0}x │", label, fmt(en), fmt(me), ratio);
}
